### PR TITLE
Order functions in the DataStore

### PR DESCRIPTION
**Pull request by tevrat-aksoy**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Bugfix
- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Build-related changes
- Documentation content changes
- Testing
- Other (please describe):

## What is the current behavior?

Resolves: #284 

## What is the new behavior?

-
-
-

### CODE REVIEW

File: src/data/data_store.cairo [hunk 1]
```diff
@@ -549,8 +571,82 @@ mod DataStore {
...
+            self.order_keys.write(count - 1, 0);
+            self.order_count.write(count - 1);
+        }
+
+        fn get_key_index(self: @ContractState, key: felt252) -> Option<u64> {
+            let mut i = 0_u64;
+            let count = self.order_count.read();
+            // Search for key index
+            let result = loop {
+                if (i >= count) {
```
Comment (enitrat) [inline src/data/data_store.cairo:620]: if i is sequentially increased by one and a precondition on the loop is that count >= 0 >= i then doing equality checks is better as range checks are more expensive

File: src/data/data_store.cairo [hunk 2]
```diff
@@ -549,8 +571,82 @@ mod DataStore {
...
+
+            let order = self.order_values.read(key);
+            // We use the zero address to indicate that the order does not exist.
+            assert(!order.account.is_zero(), DataError::ORDER_NOT_FOUND);
+
+            let index_result = self.get_key_index(key);
+            let mut index: u64 = 0;
+            match index_result {
+                Option::Some(i) => index = i,
+                Option::None(()) => assert(false, DataError::ORDER_INDEX_NOT_FOUND),
```
Comment (enitrat) [inline src/data/data_store.cairo:603]:
```suggestion
                Option::None(()) => panic_with_felt252(DataError::ORDER_INDEX_NOT_FOUND),
```

File: src/data/data_store.cairo [hunk 3]
```diff
@@ -549,8 +571,82 @@ mod DataStore {
...
+            }
+        }
+
+        fn get_order_count(self: @ContractState) -> u64 {
+            self.order_count.read()
+        }
+
+        fn get_order_keys(self: @ContractState, start: u64, end: u64) -> Array<felt252> {
+            let mut keys = ArrayTrait::<felt252>::new();
+            let mut i = start;
```
Comment (enitrat) [inline src/data/data_store.cairo:641]: we should verify that `end <order_count`

File: src/data/data_store.cairo [hunk 4]
```diff
@@ -549,8 +571,82 @@ mod DataStore {
...
+            let mut index: u64 = 0;
+            match index_result {
+                Option::Some(i) => index = i,
+                Option::None(()) => assert(false, DataError::ORDER_INDEX_NOT_FOUND),
+            };
+
+            let count = self.order_count.read();
+            let last_key = self.order_keys.read(count - 1);
+
+            // Update removed key index with last key
```
Comment (enitrat) [inline src/data/data_store.cairo:609]: Is there a requirement to preserve Orders order?

File: tests/data/test_data_store.cairo [hunk 5]
```diff
@@ -255,22 +391,41 @@ fn given_normal_conditions_when_order_functions_then_expected_results() {
...
-    // Check that the retrieved order is the same as the original order.
-    // TODO: Add a proper equality check for orders by implementing `PartialEq` for `Order`.
-    assert(retrieved_order.account == account, 'invalid order');
+/// Utility function to assert order structs
+/// This function will panic if any of the following fields do not match between the two orders:
+/// # Arguments
+///
+/// * `order1` - First order struct 
+/// * `order2` - Second order struct.
+fn assert_order_eq(order1: @Order, order2: @Order) {
```
Comment (enitrat) [inline tests/data/test_data_store.cairo:403]: I think you can just derive PartialEq on the Order type

File: src/data/data_store.cairo (hunk 1 above)
Comment (tevrat-aksoy) [reply src/data/data_store.cairo:620]: ok. I will change it

File: src/data/data_store.cairo (hunk 3 above)
Comment (tevrat-aksoy) [reply src/data/data_store.cairo:641]: yes, you are right

File: src/data/data_store.cairo (hunk 4 above)
Comment (tevrat-aksoy) [reply src/data/data_store.cairo:609]: no. I am moving key from last index to removed key index. Do you have better  idea?

Comment (enitrat) [reply src/data/data_store.cairo:609]: No I think it's a good idea

**Comment by zarboq** [comment]
Hey nice work !

Just wondering if we shouldn't follow one single standard for these functions in data_store (order, withdrawal etc.). We already have two different implementations between the one in this PR and [this one](https://github.com/keep-starknet-strange/satoru/pull/295/files#diff-abaaff167361871b5fd1badc5656ef476d6adf6f5ffde32f190dcc935d6582f5), so maybe we need to choose one.

Wdyt @abdelhamidbakhta

**Comment by enitrat** [comment]
> Actually after reading yours I had a second thought about optimisations of storage modifications. You have an extra storage variables that tracks the key associated to each index, which means that when you're removing an order you can just move the key of the last index to the removed index. 
> The solution with List<T> moves the entire Withdrawal struct, which means that instead of modifying 1 storage slot we modify N where N is the number of values stored in the withdrawals.
> 
> So it might be actually more interesting to use your solution in Withdrawals, even though it will require adding two extra storage variables, it should decrease overall gas costs

**Comment by tevrat-aksoy** [comment]
> Hey nice work !
> 
> Just wondering if we shouldn't follow one single standard for these functions in data_store (order, withdrawal etc.). We already have two different implementations between the one in this PR and [this one](https://github.com/keep-starknet-strange/satoru/pull/295/files#diff-abaaff167361871b5fd1badc5656ef476d6adf6f5ffde32f190dcc935d6582f5), so maybe we need to choose one.
> 
> Wdyt @abdelhamidbakhta
Have you made any decisions?

**Comment by zarboq** [comment]
Yeah if you're ok with eni's suggestion we can go with that !

**Comment by enitrat** [comment]
That would be for a PR to alexandria most likely 😅

**Comment by tevrat-aksoy** [comment]
I will implement with current version then. We can update List  later with alexandria update.

**Review by zarboq — APPROVED** [review]

**Pull request merged**